### PR TITLE
Add docs for ip autodetection methods

### LINKED ISSUE #494: Document new IP autodetection methods
Issue: https://github.com/projectcalico/calicoctl/issues/873
PR: https://github.com/projectcalico/calicoctl/pull/1500

### PR DESCRIPTION
**Pull request by robbrockbank**
Fixes #494

### CODE REVIEW

File: master/reference/calicoctl/commands/node/run.md [hunk 1]
```diff
@@ -104,6 +130,83 @@ CALICO_LIBNETWORK_ENABLED is true - start libnetwork service
...
+we will attempt to autodetect an IPv4 address.  An IPv6 address, however, will 
+only be autodetected when explicitly requested.
+
+To force autodetection of an IPv4 address use the option `--ip=autodetect`.  To
+force autodetection of an IPv6 address use the option `--ip6=autodetect`. 
+
+To set the autodetection method for IPv4, use the `--ip-autodetection-method` option.
+To set the autodetection method for IPv6, use the `--ip6-autodetection-method` option.
+
+>  Note, if you are starting the calico/node container directly (and not using the
```
Comment (gunjan5) [inline master/reference/calicoctl/commands/node/run.md]: we have been printing note as: 
```
> **NOTE**
> blah blah ,meep
> more blah
```
So it looks consistent in the html rendering.

File: master/reference/calicoctl/commands/node/run.md [hunk 2]
```diff
@@ -104,6 +130,83 @@ CALICO_LIBNETWORK_ENABLED is true - start libnetwork service
 Calico node started successfully
 ```
 
+#### IP Autodetection method examples
+
+The node resource includes IPv4 and IPv6 routing IP addresses that should
+match those on one of the host interfaces.  These IP addresses may be 
+configured in advance by configuring the node resource prior to starting the
+calico/node service, or alternatively the addresses may either be explicitly
```
Comment (gunjan5) [inline master/reference/calicoctl/commands/node/run.md]: `service, or alternatively the addresses` --> `service, alternatively, the addresses` ? I  think?

File: master/reference/calicoctl/commands/node/run.md [hunk 3]
```diff
@@ -104,6 +130,83 @@ CALICO_LIBNETWORK_ENABLED is true - start libnetwork service
 Calico node started successfully
 ```
 
+#### IP Autodetection method examples
+
+The node resource includes IPv4 and IPv6 routing IP addresses that should
+match those on one of the host interfaces.  These IP addresses may be 
+configured in advance by configuring the node resource prior to starting the
+calico/node service, or alternatively the addresses may either be explicitly 
+specified, or autodetected through options on the `calicoctl run` command.
```
Comment (gunjan5) [inline master/reference/calicoctl/commands/node/run.md]: get rid of the comma

File: master/reference/calicoctl/commands/node/run.md [hunk 4]
```diff
@@ -104,6 +130,83 @@ CALICO_LIBNETWORK_ENABLED is true - start libnetwork service
...
+best suited to your deployment.  If you are able to explicitly specify the IP 
+addresses, that is always preferred over autodetection. This section describes
+the available methods for autodetecting the hosts IP addresses.
+
+An IPv4 address is always required, and so if no address was previously
+configured in the node resource, and no address was specified on the CLI, then
+we will attempt to autodetect an IPv4 address.  An IPv6 address, however, will 
+only be autodetected when explicitly requested.
+
+To force autodetection of an IPv4 address use the option `--ip=autodetect`.  To
```
Comment (gunjan5) [inline master/reference/calicoctl/commands/node/run.md]: `IPv4 address use` --> `IPv4 address, use`

File: master/reference/calicoctl/commands/node/run.md [hunk 5]
```diff
@@ -104,6 +130,83 @@ CALICO_LIBNETWORK_ENABLED is true - start libnetwork service
...
+force autodetection of an IPv6 address use the option `--ip6=autodetect`.
```
Comment (gunjan5) [inline master/reference/calicoctl/commands/node/run.md]: address`,` use

File: master/reference/calicoctl/commands/node/run.md [hunk 6]
```diff
@@ -104,6 +130,83 @@ CALICO_LIBNETWORK_ENABLED is true - start libnetwork service
...
+To set the autodetection method for IPv6, use the `--ip6-autodetection-method` option.
+
+>  Note, if you are starting the calico/node container directly (and not using the
+>  `calicoctl run` helper command), the options are passed in an environment 
+>  variables.  These are described in the [calico/node configuration guide]({{site.baseurl}}/{{page.version}}/reference/node/configuration)).
+
+**first-found**
+
+The `first-found` option enumerates all interface IP addresses and returns the
+first valid IP address (best guess based on IP version and type of address) on
```
Comment (gunjan5) [inline master/reference/calicoctl/commands/node/run.md]: best guess? Can this sentence be reworded? Not sure I understand what it means

File: master/reference/node/configuration.md [hunk 7]
```diff
@@ -27,3 +29,54 @@ The `calico/node` container is primarily configured through environment variable
...
+The `interface` method uses the supplied interface regular expression (golang
+syntax) to enumerate matching interfaces and to return the first IP address on
+the first matching interface.  Note that order that interfaces are listed is
+system dependent.
+
+e.g.
+```
+# Valid IP address on interface eth0
+IP_AUTODETECT_METHOD=interface=eth0
+IP6_AUTODETECT_METHOD=interface=eth0
```
Comment (gunjan5) [inline master/reference/node/configuration.md]: can we maybe add an example regex here? So users know if they need to put quotes or wherever around the regex while passing it to the variable

File: master/reference/node/configuration.md [hunk 8]
```diff
@@ -14,6 +14,8 @@ The `calico/node` container is primarily configured through environment variable
...
 | IP6 | The IPv6 address for Calico will bind to. When specified, the address is saved in the  [node resource configuration]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/node) for this host, overriding any previously configured value. When omitted, if an address has not yet been configured in the node resource, IPv6 routing is not enabled. When omitted, if an IPv6 address has been previously configured in the node resource, IPv6 is enabled using the already configured address. | ipv6 address | |
+| IP_AUTODETECT_METHOD| The method to use to autodetect the IPv4 address for this host. This is only used when the IPv4 address is being autodetected. See section below for details of the valid methods. | string | first-found |
```
Comment (gunjan5) [inline master/reference/node/configuration.md]: maybe drop a line in the second column saying. See #IPAutodetectionmethods hyperlink for more options?

File: master/reference/calicoctl/commands/node/run.md (hunk 1 above)
Comment (robbrockbank) [inline master/reference/calicoctl/commands/node/run.md]: awesome!  didn't know that, thanks!

**Pull request merged**